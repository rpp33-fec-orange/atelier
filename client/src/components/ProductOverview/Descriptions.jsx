import React from 'react';

const Descriptions = (props) => (

  <div id="descriptions">
    <h4>Slogan</h4>
    Description
  </div >
)

export default Descriptions;