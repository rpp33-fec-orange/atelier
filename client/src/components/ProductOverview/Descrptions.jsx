import React from 'react';

class Descriptions extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      product: {}
    }
  }

  render() {
    return (
      <div id="descriptions">
        Esconced yourself in the finest wool in all of Middle Earth, crafted with Elven Everlast threads with flax fiber from the Shire with a dash of gunpowder of Modor.<br></br>
      </div >
    )
  }
}

export default Descriptions;