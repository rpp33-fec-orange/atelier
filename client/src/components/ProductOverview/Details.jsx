import React from 'react';
import $ from 'jquery';

//Object.keys(props.productStylesById.results.skus).map

const Details = (props) => (
  <div id="details">
    <div id="rating">★★★★☆</div>
    <div id="category">{props.productById.category}</div>
    <div id="name">{props.productById.name}</div>
    <div id="price">{props.productById.default_price}</div>
    <div id="selector">Select Style/Size/Quantity</div>
    <select id="style">
      {props.productStylesById.results.map((singleData) =>
        <option value={singleData.name}>{singleData.name}</option>
      )}
    </select>
    <select id="size">
      <option value="xs">XS</option>
      <option value="s">S</option>
      <option value="m">M</option>
      <option value="l">L</option>
    </select>
    <select id="quantity">
      <option value="one">1</option>
      <option value="two">2</option>
      <option value="three">3</option>
      <option value="four">4</option>
      <option value="four">5</option>
    </select><br></br>
    <button id="add">ADD TO CART</button><button id="favorte">☆</button>
  </div >
)

// class Details extends React.Component {
//   constructor(props) {
//     super(props);
//     this.state = {
//       product: {}
//     }
//   }

//   render() {
//     return (
//       <div id="details">
//         <div id="rating">★★★★☆</div>
//         <div id="category">{this.state.product.category}</div>
//         <div id="name">{this.state.product.name}</div>
//         <div id="price">{this.state.product.default_price}</div>
//         <div id="selector">Select Style/Size/Quantity</div>
//         <select id="style">
//           <option value="gray">Heather Gray</option>
//           <option value="black">Graphite Black</option>
//           <option value="white">Moonlight White</option>
//         </select>
//         <select id="size">
//           <option value="small">Small</option>
//           <option value="medium">Medium</option>
//           <option value="large">Large</option>
//           <option value="extra-large">Extra Large</option>
//         </select>
//         <select id="quantity">
//           <option value="one">1</option>
//           <option value="two">2</option>
//           <option value="three">3</option>
//           <option value="four">4</option>
//         </select><br></br>
//         <button id="add">ADD TO CART</button><button id="favorte">☆</button>
//       </div >
//     )
//   }
// }

export default Details;