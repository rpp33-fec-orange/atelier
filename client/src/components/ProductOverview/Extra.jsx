import React from 'react';

class Extra extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      product: {}
    }
  }

  render() {
    return (
      <div id="extra">
        ✔ Range Free Unicorn Hair <br></br>
        ✔ Twice Reborn Phoenix Feather <br></br>
        ✔ Earth Core Crystalized Lava
      </div >
    )
  }
}

export default Extra;