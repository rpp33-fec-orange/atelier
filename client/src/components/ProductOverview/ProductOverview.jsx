import React from 'react';

const ProductOverview = (props) => (
  <div>
    <h4 id="productOverview">Product Overview</h4>
  </div>
)

export default ProductOverview;