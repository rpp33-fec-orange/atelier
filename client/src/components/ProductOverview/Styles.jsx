import React from 'react';
import $ from 'jquery';

class Styles extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      productById: props.productById,
      rating: '★★★★☆',
      productStylesById: props.productStylesById,
      styles: props.productStylesById.results,
      currentStyle: props.productStylesById.results[0],
      currentStyleSkus: props.productStylesById.results[0].skus,
      currentSku: props.productStylesById.results[0].skus[Object.keys(props.productStylesById.results[0].skus)[0]],
      skuCode: '',
      mainPhotoURL: props.productStylesById.results[0].photos[0].url,
      subPhotos: props.productStylesById.results[0].photos,
      quantityArray: [],
      quantitySelected: 0,
      cart: []
    }
    this.photoClick = this.photoClick.bind(this);
    this.styleChange = this.styleChange.bind(this);
    this.skuChange = this.skuChange.bind(this);
    this.postCart = this.postCart.bind(this);
    this.getCart = this.getCart.bind(this);
    this.favoriteClick = this.favoriteClick.bind(this);
    this.makeQuantityArray = this.makeQuantityArray.bind(this);
    this.quantityChange = this.quantityChange.bind(this);
    this.readReviews = this.readReviews.bind(this);
  }

  photoClick(e) {
    this.setState({
      mainPhotoURL: e.target.src
    })
  }

  styleChange(e) {
    for (let i = 0; i < this.state.styles.length; i++) {
      if (this.state.styles[i].name === e.target.id) {
        let selectedStyle = this.state.styles[i];
        this.setState({
          currentStyle: selectedStyle,
          currentStyleSkus: selectedStyle.skus,
          mainPhotoURL: selectedStyle.photos[0].url,
          subPhotos: selectedStyle.photos
        });
      }
    }
  }

  skuChange(e) {
    let skuKeys = Object.keys(this.state.currentStyleSkus);
    for (let i = 0; i < skuKeys.length; i++) {
      if (this.state.currentStyleSkus[skuKeys[i]].size === e.target.value) {
        let selectedSku = this.state.currentStyleSkus[skuKeys[i]];
        this.setState({
          currentSku: selectedSku,
          skuCode: skuKeys[i]
        });
      }
    }
    this.makeQuantityArray();
  }

  postCart() {
    let cartItem = {
      sku: this.state.skuCode,
      quantity: this.state.quantitySelected
    };
    this.state.cart.push(cartItem);
    $.ajax({
      context: this,
      type: 'POST',
      url: '/cart',
      data: JSON.stringify({ cartItem }),
      contentType: 'application/json',
      success: function (successAjax) {
        console.log('Ajax POST Success!');
      },
      error: function (errorAjax) {
        console.log('Ajax POST Error!');
      },
    })
  }

  getCart() {
    console.log('html cart: ', this.state.cart);
    $.ajax({
      context: this,
      type: 'GET',
      url: '/cart',
      success: function (success) {
        console.log('getCart ajax GET success: ', success);
        this.setState({
          cart: success,
        })
      },
      error: function (error) {
        console.log('getCart ajax GET error: ', error);
      },
      contentType: "application/json",
    })
    console.log('api cart: ', this.state.cart);
  }

  favoriteClick() {

  }

  readReviews() {

  }

  makeQuantityArray() {
    let updatedLength = this.state.currentSku.quantity + 1;
    var arrayCreated = Array.from({ length: updatedLength }, (v, i) => i);
    let shiftedArray = arrayCreated.shift();
    this.setState({
      quantityArray: arrayCreated
    })
  }

  quantityChange(e) {
    this.setState({
      quantitySelected: e.target.value
    })
  }

  render() {
    let productById = this.state.productById;
    let productStylesById = this.state.productStylesById;
    let rating = this.state.rating;
    let styles = this.state.styles;
    let currentStyle = this.state.currentStyle;
    let currentStyleSkus = this.state.currentStyleSkus;
    let currentSku = this.state.currentSku;
    let mainPhotoURL = this.state.mainPhotoURL;
    let subPhotos = this.state.subPhotos;
    let quantityArray = this.state.quantityArray;
    console.log('product styles', productStylesById);
    return (
      <div>
        <div id="photos">
          <img id="mainPhoto" src={mainPhotoURL} width="300" height="425"></img> <br></br>
          {subPhotos.map((photo) =>
            <img id="subPhoto" src={photo.url} width="75" height="105" onClick={this.photoClick}></img>
          )}
        </div>
        <div id="styles">
          <div id="rating">{rating} <button onClick={this.readReviews}>Read Reviews</button></div>
          <div id="category">{productById.category}</div>
          <div id="name">{productById.name}</div>
          <div id="price">{productById.default_price}</div>
          <div data-testid="selector" id="selector">Select Style/Size/Quantity</div>
          <div id="style">
            {styles.map((style) =>
              <img id={style.name} src={style.photos[0].url} width="35" height="45" onClick={this.styleChange}></img>
            )}
          </div>
          <select id="size" onChange={this.skuChange}>
            <option value="nullSize">Select Size</option>
            {Object.keys(currentStyleSkus).map((sku) =>
              <option value={currentStyleSkus[sku].size}>{currentStyleSkus[sku].size}</option>
            )}
          </select>
          <select id="quantity" onChange={this.quantityChange}>
            {quantityArray.map((quantityItem) =>
              <option value={quantityItem}>{quantityItem}</option>
            )}
          </select><br></br>
          {currentSku.quantity ? <button id="postCart" onClick={this.postCart}>ADD TO CART</button> : <button id="outOfStock" disabled>Out of Stock</button>}
          <button id="favorite" onClick={this.favoriteClick}>☆</button><button id="getCart" onClick={this.getCart}>YOUR CART</button>
        </div >
      </div >
    )
  }
}

export default Styles;