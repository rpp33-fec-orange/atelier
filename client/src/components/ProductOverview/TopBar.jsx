import React from 'react';

class TopBar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      keyword: ''
    }
    this.onChange = this.onChange.bind(this);
    this.clickHandler = this.clickHandler.bind(this);
  }

  onChange(e) {
    this.setState({
      keyword: e.target.value
    });
  }

  clickHandler() {
    this.props.searchHandler(this.state.keyword);
    this.setState({
      keyword: ''
    });
  }

  render() {
    return (
      <div id="TopBar">
        <div id="logoSearch">
          <div data-testid="logo" id="logo">Atelier<input id="input" placeholder="keyword" onChange={this.onChange}></input><button id="inputButton" onClick={this.clickHandler}>Search</button></div>
        </div>
      </div >
    )
  }
}

export default TopBar;