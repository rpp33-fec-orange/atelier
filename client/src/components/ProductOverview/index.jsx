import React from 'react';
import $ from 'jquery';
import TopBar from './TopBar.jsx';
import Styles from './Styles.jsx';
import Info from './Info.jsx';

class ProductOverview extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      id: props.id,
      productById: {},
      productStylesById: {},
      styles: [],
      initialized: false
    }
    this.productHandler = this.productHandler.bind(this);
    this.stylesHandler = this.stylesHandler.bind(this);
    this.searchHandler = this.searchHandler.bind(this);
  }

  productHandler() {
    $.ajax({
      context: this,
      type: 'GET',
      url: `/products/${this.state.id}`,
      success: function (success) {
        console.log('product overview productHandler ajax GET success');
        this.setState({
          productById: success,
        })
      },
      error: function (error) {
        console.log('product overviewnproductHandler ajax GET error: ', error);
      },
      contentType: "application/json",
    })
  }

  stylesHandler() {
    $.ajax({
      context: this,
      type: 'GET',
      url: `/products/${this.state.id}/styles`,
      success: function (success) {
        console.log('product overview productHandler ajax GET success');
        this.setState({
          productStylesById: success,
          styles: success.results,
          initialized: true
        })
      },
      error: function (error) {
        console.log('product overview productHandler ajax GET error: ', error);
      },
      contentType: "application/json",
    })
  }

  searchHandler(keyword) {
    $.ajax({
      context: this,
      type: 'POST',
      url: '/search',
      data: JSON.stringify({ keyword }),
      success: function (success) {
        console.log('product overview searchHandler ajax POST success');
      },
      error: function (error) {
        console.log('product overview searchHandler ajax POST error: ', error);
      },
      contentType: "application/json",
    })
    alert(`${keyword} was searched!`);
  }

  componentDidMount() {
    this.productHandler();
    this.stylesHandler();
  }

  componentDidUpdate(prevProps) {
    if (this.props.id !== prevProps.id) {
      this.setState({
        id: this.props.id,
      })
    }
    this.productHandler();
    this.stylesHandler()
    // this.searchHandler();
  }

  render() {
    if (this.state.initialized) {
      return (
        <div class="index-container" id="productOverview">
          <div class="index-item index-item-1">
            <TopBar data-testid="topbar?" searchHandler={this.searchHandler} />
          </div>
          <div id="message">Hack Reactor Sale! ---Extra 40% off on selected items--- Free shipping for orders over $50!</div>
          <div class="index-item index-item-2">
            <Styles data-testid="styles?" productById={this.state.productById} productStylesById={this.state.productStylesById} rating={this.props.rating} currentStyleHandler={this.props.currentStyleHandler} yourOutfitHandleClick={this.props.yourOutfitHandleClick} />
          </div>
          <div class="index-item index-item-2">
            <Info data-testid="descriptions?" productById={this.state.productById} />
          </div>
        </div>
      )
    } else {
      return (
        <div data-testid="loading" id="loading">
          â‡† Loading...
        </div>
      )
    }
  }
}

export default ProductOverview;