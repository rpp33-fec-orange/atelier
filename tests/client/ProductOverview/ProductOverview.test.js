import React from 'react';
import TestRenderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { shallow } from 'enzyme';
import { render, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom'
import ProductOverview from '../../../client/src/components/ProductOverview/index.jsx';
import Descriptions from '../../../client/src/components/ProductOverview/Descriptions.jsx';
import Styles from '../../../client/src/components/ProductOverview/Styles.jsx';
import TopBar from '../../../client/src/components/ProductOverview/TopBar.jsx';

Enzyme.configure({ adapter: new Adapter() });

describe('ProductOverview component tests', () => {
  const testRenderer = TestRenderer.create(<ProductOverview />);
  const testInstance = TestRenderer.root;

  test('it should contain a h4 element with the text "⇆ Loading..." in index.jsx', () => {
    expect(testInstance.findByProps({ id: "loading" }).children).toEqual(['⇆ Loading...']);
  });

  test('render loading page on index.jsx on start up', () => {
    render(<ProductOverview />);
    const element = screen.getByTestId('loading');
    expect(element).toBeInTheDocument();
    expect(element).toHaveTextContent('⇆ Loading...');
  });

  test('render logo at TopBar component startup', () => {
    render(<TopBar />);
    const element = screen.getByTestId('logo');
    expect(element).toBeInTheDocument();
    expect(element).toHaveTextContent('Atelier');
  });

  test('matches snapshot ', () => {
    const tree = TestRenderer.create(<ProductOverview />).toJSON();
    expect(tree).toMatchSnapshot();
    console.log(tree);
  });

  it("should render my component", () => {
    const component = shallow(<ProductOverview />);
    expect(component.exists()).toBe(true);
  });

  it("should render my component", () => {
    const component = shallow(<Descriptions />);
    expect(component.exists()).toBe(true);
  });

  it("should render my component", () => {
    const component = shallow(<Styles />);
    expect(component.exists()).toBe(true);
  });

  it("should render my component", () => {
    const component = shallow(<TopBar />);
    expect(component.exists()).toBe(true);
  });
});