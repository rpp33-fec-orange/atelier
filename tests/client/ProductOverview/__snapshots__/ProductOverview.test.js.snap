// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductOverview component tests it should match snapshot  1`] = `
<div
  data-testid="loading"
  id="loading"
>
  â‡† Loading...
</div>
`;
